Templatize and cache compiled BPF programs

joestringer opened the pull request:
Based on the following PRs:
* [x] #7197 - testutils package fixes to allow unit tests to work
* [x] #7180 - Make map garbage collection unit-testable
* [x] #7196 - Introduce ELF library
* [x] #7218 - Make map names static-length
* [x] #7219 - Refactor and simplify some datapath compile/configuration code

Cilium is designed to tailor the datapath forwarding logic for each individual endpoint, minimizing the amount of code necessary to implement the desired behaviour: endpoint-specific data is compiled directly into the machine-code instructions; unused features are compiled out; policies are implemented local to the endpoint as O(1) lookups.

Up until now, to achieve these goals, Cilium has required at least one invocation of the C compiler per endpoint run.

No more.

This PR templatizes the BPF datapath code, allowing the daemon to compile the datapath just once then inject endpoint-specific object references into the ELF files after compilation, which is about 25x faster (Same order of magnitude as simply reloading the datapath):

```
BenchmarkCompileAndLoad-4             10         157922809 ns/op
BenchmarkCompileOrLoad-4             200           6989403 ns/op
BenchmarkReplaceDatapath-4           200           5534560 ns/op
```

This is achieved broadly through the following set of changes in this PR:
1. Modify the ELF format used by Cilium to generate symbols for static data.
1. Introduce an ELF package which allows parsing and rewriting symbols for objects on the filesystem.
1. Split endpoint-specific data from general datapath configuration data.
1. Hash all header files for general datapath configuration data and code.
1. Build a cache of datapath configurations that have been created before based on the hash, and store the corresponding ELF objects on the filesystem in a canonical location.
   * When a template ELF is available, use the ELF library to copy & rewrite endpoint-specific data into the endpoint directory
   * The cache is written such that hashing of the general datapath configuration occurs first, then the builder checks whether a channel is closed (indicating that the template is already built). If not, create a new channel for others to block on and start compiling. When done, close the channel to indicate compilation is complete.
   * If the build fails, or if another thread is building which fails, the channel will be closed and the current goroutine will attempt to fetch the resulting build and fail. This would return an error, causing the endpoint to go into `not-ready` state just like today.

Commits are arranged for review of individual patches.

Primary potential upgrade impact that may have some unintended effects is the rename of endpoint maps. To ease ELF variable/string substitution, I have rearranged all endpoint-specific map names to be in the form `%s_%05d`, eg `my_map_00042`. Maps with the old form name `my_map_42` will be deleted on startup. If we need to make this less disruptive, then we can look at that; if you're aware of something important in this area then please raise it during review.

Fixes: https://github.com/cilium/cilium/issues/6618

Tasks:
* [x] Fix up MAC addresses
* [x] Solve stale identity problem (identity `5` is written into ELF files in docker env)
* [x] Fix LB test failure
* [x] Fix CIDR policy issues
* [x] Address BPF complexity limit restrictions
* [x] Update the Cilium VM/Docker images
  * #7114 
  * https://github.com/cilium/packer-ci-build/pull/119
  * ~Consider the inline-asm approach instead for iproute2~
    * inline-asm requires additional refactor of datapath code, skipping for now
  * Based on https://github.com/cilium/iproute2/tree/static-data
    * PPA: https://launchpad.net/~joestringer/+archive/ubuntu/ppa
    * OBS/Tumbleweed: https://build.opensuse.org/package/show/security:netfilter/iproute2

* [x] GC old template binaries
* [x] Address impact of endpoint-specific map renames
* [x] Tidy for review

Future work:
* CLI, template links on filesystem
* Restoring templates
* For more, see https://github.com/cilium/cilium/projects/34

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/7095)
<!-- Reviewable:end -->

joestringer committed:
bpf: Replace data definitions with static variables

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
datapath: Parameterize endpoint static data

Factor out static data write and add tests for it, against the ELF
libraries.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
Add lib directory to CFLAGS includes

This will be necessary for an upcoming commit to successfully compile
template programs, otherwise the endpoint and node config header files
will not be able to pick up the `bpf/lib/utils.h` path correctly.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
loader: Add benchmark for noop ELF write

Add baseline benchmark for writing an existing elf to a new location.

Example run:

    $ sudo make bench-privileged TESTPKGS=pkg/datapath/loader BENCH=CompileOrLoad
    pkg: github.com/cilium/cilium/pkg/datapath/loader
    BenchmarkCompileOrLoad-4             200          10715695 ns/op
    PASS
    ok      github.com/cilium/cilium/pkg/datapath/loader    3.494s

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
endpoint: Flesh out endpoint cache for ELF reload

Cache everything that will be necessary for upcoming access by the
loader while not holding the endpoint lock.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
loader: Cache datapath objects

* When attempting to CompileAndLoad progs for an endpoint, first
  hash the endpoint header config and see if there's a template.
* If there's no such template, generate a new template header and
  objectfile, then cache this.
* Otherwise if there's a template, use it!

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
loader: Support metrics for template compilation

The recent commit ("loader: Cache datapath objects") reduced the set
of situations where a datapath recompilation is necessary, but retained
the existing bpfCompilation metric in place. This causes ELF rewrites to
be classified as compiles, which is strictly wrong.

Fix this up by introducing a new spanstat in the loader package and
embedding that in the endpoint's statistics object, then only accounting
the stats when template compilation occurs.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
bpf: Fix up default config files for template builds

Update the bpf/*_config.h files for the upcoming templating work.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
loader: Garbage collect template directory on startup

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
loader: Introduce datapath templates

Introduce template wrappers around an endpoint. This passes through all
datapath configuration parameters that must be applied to implement the
current configuration of the endpoint, but substitutes all static data
for dummy values (typically 0xFFFF).

Add tests for the other side of the logic which substitutes the template
values back into the endpoint's actual static data. This information
will be passed to the ELF library by upcoming loader logic for
substitution in ELF.Write().

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
loader: Hash datapath objects and store results

Construct a cache of hashes based on the current datapath code (SHA1SUM
of the bpf/ directory plus generated headerfiles). This will be used by
upcoming commits first to replace the endpoint headerfile hashing from
pkg/endpoint, then subsequently to cache templated datapaths and avoid
compiling if an endpoint can be generated from a previously compiled
template object.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer commented on pkg/elf/symbols.go:
```diff
@@ -0,0 +1,247 @@
+	dataSection       = ".data"
+	mapSection        = "maps"
+	endpointMapPrefix = "1/0x"     // CILIUM_MAP_POLICY ## / ## LXC_ID
+	endpointRelPrefix = ".rel1/0x" // .rel ## endpointMapPrefix
```
nit: this PR adds `CallString()` for this, should ideally share it.

joestringer commented on pkg/elf/elf.go:
```diff
@@ -0,0 +1,260 @@
+// On success, writes the file to the specified path. On failure, returns an
+// error and no file is left on the filesystem.
+//
+// Concurrent writes from the same source ELF is NOT safe.
```
nit: while concurrent write from the same ELF struct is not safe, concurrent copy from the same source object file should be fine. Workaround is to Open() the same object multiple times.

joestringer commented on pkg/datapath/loader/metrics.go:
```diff
@@ -0,0 +1,32 @@
+
+// SpanStat is a statistics structure for storing metrics related to datapath
+// load operations.
+type SpanStat struct {
```
nit: We don't measure datapath load (elf rewrite + load) here, but that can be an easy subsequent patch.

joestringer commented on pkg/endpoint/bpf.go:
```diff
@@ -61,7 +58,7 @@ func (e *Endpoint) PolicyMapPathLocked() string {
 // CallsMapPathLocked returns the path to cilium tail calls map of an endpoint.
 func (e *Endpoint) CallsMapPathLocked() string {
-	return bpf.LocalMapPath(CallsMapName, e.ID)
+	return bpf.LocalMapPath(loader.CallsMapName, e.ID)
```
nit: `pkg/datapath/loader` is not the most natural place for this, but it worked without introducing import cycles. A subsequent patch could clean this up to put it maybe in `pkg/datapath/maps` or move it with the templates stuff into a `pkg/datapath/template` package, but additional refactoring is required and I wanted to get this out for review of the core functionality first.

joestringer commented on bpf/node_config.h:
```diff
@@ -26,11 +26,12 @@
 #define ENCAP_IFINDEX 1
 #define HOST_IFINDEX 1
-#define HOST_IP 0xbe, 0xef, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xa, 0x0, 0x2, 0xf, 0xff, 0xff
+DEFINE_IPV6(HOST_IP, 0xbe, 0xef, 0x0, 0x0, 0x0, 0x0, 0x0, 0x1, 0x0, 0x0, 0xa, 0x0, 0x2, 0xf, 0xff, 0xff);
```
These are reworked just so that the IPV6 accessors can be consistent across the codebase, not because we're templating these values (they're not expected to change anyway).

joestringer committed:
elf: Add IPv6 definition unit tests

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
serializer: Add support for waiting for close

Add a Wait() function which returns when either the FunctionQueue is
closed, or a context deadline is exceeded.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
endpoint: Use pkg/datapath/cache for header hash

Replace the headerfile read from the filesystem and corresponding hash
code with a call into the new datapath hash cache library.

Signed-off-by: Joe Stringer <joe@cilium.io>

ianvernon commented on pkg/datapath/maps/map.go:
```diff
@@ -95,9 +95,14 @@ func newMapSweeper(g endpointManager) *mapSweeper {
-		if !ms.endpointExists(epID) {
+		if ms.endpointExists(epID) {
+			prefix := strings.TrimSuffix(filename, endpointID)
+			if filename != bpf.LocalMapName(prefix, uint16(tmp)) {
```
why not use epID instead of doing `uint16(tmp)` again?

joestringer replied:
;) Probably because this code was shuffled around a couple of times by the time the dependent PRs were merged. Will fix.

ianvernon commented on pkg/datapath/loader/template.go:
```diff
@@ -0,0 +1,194 @@
+}
+
+// sliceToU16 converts the input slice of two bytes to a uint16.
+func sliceToU16(input []byte) uint16 {
```
The code will panic if the provided slice is of length <=1 - maybe guard against this? Not sure how necessary it is given that this function is not exported. Also maybe guard against case where length > 2 ?

joestringer replied:
I agree if this were a library that we're trying to support arbitrary inputs, but here we're using it in a very constrained way with known types - specifically the MAC type, which the caller takes 2-byte slices to pass in. There's no point in adding runtime checks when this is well-defined. (If we refactor this somewhere for reuse in future, I agree we should add sanity checks like this).

ianvernon commented on pkg/datapath/loader/template.go:
```diff
@@ -0,0 +1,194 @@
+}
+
+// sliceToBe16 converts the input slice of two bytes to a big-endian uint16.
+func sliceToBe16(input []byte) uint16 {
```
Same here - enforce that byte slice is exactly 2 bytes ?

joestringer replied:
Ditto to above.

ianvernon commented on pkg/datapath/loader/cache.go:
```diff
@@ -65,6 +65,17 @@ func Init(dp datapath.Datapath, nodeCfg *datapath.LocalNodeConfiguration) {
 
+// RestoreTemplates populates the object cache from templates on the filesystem
+// at the specified path.
+func RestoreTemplates(stateDir string) error {
```
Is there a plan to restore these eventually?

joestringer replied:
Yep, that's why it's called RestoreTemplates. I figured this is unnecessary to get the basics working, but it should be a trivial patch on top. Will defer for now, but there's an entry in the project for this.

ianvernon replied:
Awesome :D

ianvernon commented on pkg/datapath/loader/hash.go:
```diff
@@ -0,0 +1,99 @@
+	return h
+}
+
+// Copy the cached mash of the hashed datapath paths.
```
:D

ianvernon reviewed (approved):
Nothing blocking here - great work!!!

joestringer committed:
datapath/maps: Reuse variable

A couple of lines up, uint16(tmp) is determined, should reuse the local
variable. No functional change.

Signed-off-by: Joe Stringer <joe@cilium.io>

aanm commented on pkg/datapath/loader/cache.go:
```diff
@@ -72,6 +109,111 @@ func (o *ObjectCache) Update(nodeCfg *datapath.LocalNodeConfiguration) {
+
+func (o *ObjectCache) lookup(hash string) (string, bool) {
+	o.Lock()
+	defer o.Unlock()
```
remove the defer and add the unlock on line 130-131

joestringer replied:
Can you explain why? I don't understand the difference. I mean perhaps logically it is adding a function to a list to call at the end of the function, but this seems also like something that if there's a super simple case like this, the compiler can just reorder/copy the defer function call so it happens before return. Especially when there's only one exit to the function like this.

aanm commented on pkg/datapath/loader/cache.go:
```diff
@@ -72,6 +109,111 @@ func (o *ObjectCache) Update(nodeCfg *datapath.LocalNodeConfiguration) {
+
+func (o *ObjectCache) insert(hash, objectPath string) {
+	o.Lock()
+	defer o.Unlock()
```
same here

aanm replied:
Now I'm curious if how munch you gain without the defer here. Did you write any benchmarks for this so we can compare it?

joestringer replied:
There's the `BenchmarkCompileOrLoad()`.

For what it's worth, I changed all the defer stuff and it wasn't clear to me that there was any meaningful difference. Subsequent runs on my local laptop range from about 5-8ms/op. Obviously not a rigorous evaluation but enough to tell me that there's no easy gains here.

Example:
```
$ sudo make bench-privileged TESTPKGS=pkg/datapath/loader BENCH=CompileOrLoad
can't load package: package github.com/cilium/cilium/test/bpf: C source files not allowed when not using cgo or SWIG: bpf-event-test.c elf-demo.c unit-test.c
go test -ldflags "-X github.com/cilium/cilium/pkg/kvstore.consulDummyConfigFile=/tmp/cilium-consul-certs/cilium-consul.yaml -X github.com/cilium/cilium/pkg/testutils.CiliumRootDir=/home/joe/work/src/github.com/cilium/cilium -X github.com/cilium/cilium/pkg/datapath.DatapathSHA=1234567890abcdef7890" -test.v -timeout 360s -bench=CompileOrLoad -run=^-benchtime=10s -tags=privileged_tests github.com/cilium/cilium/pkg/datapath/loader || exit 1;
goos: linux
goarch: amd64
pkg: github.com/cilium/cilium/pkg/datapath/loader
BenchmarkCompileOrLoad-4             300           5900681 ns/op
PASS
ok      github.com/cilium/cilium/pkg/datapath/loader    2.691s
```

Incremental patch diff for reference:
```
$ git di
diff --git a/pkg/datapath/loader/cache.go b/pkg/datapath/loader/cache.go
index 7f99fc285bdd..905c6b9ec210 100644
--- a/pkg/datapath/loader/cache.go
+++ b/pkg/datapath/loader/cache.go
@@ -116,8 +116,8 @@ func (o *ObjectCache) Update(nodeCfg *datapath.LocalNodeConfiguration) {
        newHash := HashDatapath(o.Datapath, nodeCfg, nil, nil)
 
        o.Lock()
-       defer o.Unlock()
        o.baseHash = newHash
+       o.Unlock()
 }
 
 // serialize finds the channel that serializes builds against the same hash.
@@ -125,27 +125,26 @@ func (o *ObjectCache) Update(nodeCfg *datapath.LocalNodeConfiguration) {
 // datapath for this hash.
 func (o *ObjectCache) serialize(hash string) (fq *serializer.FunctionQueue, found bool) {
        o.Lock()
-       defer o.Unlock()
-
        fq, compiled := o.compileQueue[hash]
        if !compiled {
                fq = serializer.NewFunctionQueue(1)
                o.compileQueue[hash] = fq
        }
+       o.Unlock()
        return fq, compiled
 }
 
 func (o *ObjectCache) lookup(hash string) (string, bool) {
        o.Lock()
-       defer o.Unlock()
        path, exists := o.toPath[hash]
+       o.Unlock()
        return path, exists
 }
 
 func (o *ObjectCache) insert(hash, objectPath string) {
        o.Lock()
-       defer o.Unlock()
        o.toPath[hash] = objectPath
+       o.Unlock()
 }
 
 // build attempts to compile and cache a datapath template object file
@@ -207,9 +206,10 @@ func (o *ObjectCache) FetchOrCompile(ctx context.Context, cfg datapath.EndpointC
        fq, compiled := o.serialize(hash)
        if !compiled {
                fq.Enqueue(func() error {
-                       defer fq.Stop()
                        templateCfg := wrap(cfg, stats)
-                       return o.build(ctx, templateCfg, hash)
+                       err := o.build(ctx, templateCfg, hash)
+                       fq.Stop()
+                       return err
                }, serializer.NoRetry)
        }
 
```

Honestly the above looks like something that the compiler could trivially optimize.

joestringer replied:
One more note on insert here, this is rare - this requires completely different datapath configuration. If this is hit often, then the entire feature is broken. We are only inserting _template_ object paths into this cache, not the actual _endpoint_ objects. (Just realising this may not be clear from my naming of variables, apologies about that).

joestringer commented on pkg/datapath/loader/cache.go:
```diff
@@ -0,0 +1,235 @@
+//
+// Returns the path to the compiled template datapath object and whether the
+// object was compiled, or an error.
+func (o *ObjectCache) FetchOrCompile(ctx context.Context, cfg datapath.EndpointConfiguration, stats *SpanStat) (string, bool, error) {
```
We actually end up taking the lock once to serialize here, then once more to look up the result once we know the template is compiled. This could be folded into one lock in the "template already exists" case. But that said there's probably not going to be a lot of contention on this, and it's not run super often either.

aanm commented on pkg/datapath/loader/cache.go:
```diff
@@ -186,15 +189,15 @@ func (o *ObjectCache) build(ctx context.Context, cfg *templateCfg, hash string)
 // Returns the path to the compiled template datapath object and whether the
 // object was compiled, or an error.
-func (o *ObjectCache) FetchOrCompile(ctx context.Context, cfg datapath.EndpointConfiguration) (string, bool, error) {
+func (o *ObjectCache) FetchOrCompile(ctx context.Context, cfg datapath.EndpointConfiguration, stats *SpanStat) (string, bool, error) {
```
Could we gain anything if `SpanStat` was an interface?

joestringer replied:
I don't think so, there's only one user of this and we don't want to inject dependencies here. Earlier this was being called from out-of-package, but not any more. This function can now be hidden.

aanm commented on pkg/datapath/loader/cache.go:
```diff
@@ -0,0 +1,79 @@
+	newHash := HashDatapath(o.Datapath, nodeCfg, nil, nil)
+
+	o.Lock()
+	defer o.Unlock()
```
same here for the defer

aanm commented on pkg/datapath/loader/hash.go:
```diff
@@ -0,0 +1,99 @@
+// Copy the cached mash of the hashed datapath paths.
+func (h *Hash) Copy() *Hash {
+	newHash := sha1.New()
+	newHash.Write(h.Data())
```
This does not create a Copy, this is creating a new Hash with a different sum!

joestringer replied:
Sigh, true. The base hash libraries didn't seem to do what I wanted them to do so this is basically an attempt to hack a copy operation in. Really it's a `New()` operation that derives from the same hash, which was deterministic enough at least to get the basic template hashing to work consistently -- ie the sum is the same between multiple endpoints on which you call `baseHash.SumEndpoint(...)` with the same input.

Do you have any suggestions for a saveable hash structure library where you don't have to do either of the following?
* Share the same hash object (lock and contend with other goroutines)
* Rehash the same base data again and again

Ideally we could also seed it given that we have already performed the first portion of the hashing all the way back in the build process with bpf.sha.

joestringer replied:
OK, I was missing that with `encoding.BinaryMarshaler`, standard library hashes can be serialized then unserialized to a new value and the new value returned, which is equivalent to making a copy.

aanm commented on pkg/serializer/func_queue.go:
```diff
@@ -81,6 +85,18 @@ func (fq *FunctionQueue) Stop() {
+// was completed.
+func (fq *FunctionQueue) Wait(ctx context.Context) error {
+	select {
+	case <-fq.stopCh:
```
fun fact of the day, if both fq.stopCh and ctx.Done are closed at the exact same time, golang will randomly pick one of these 2 cases, so to be precise you might want to return `ctx.Err()` on line 97 to ensure that in case both channels got closed at the same time and the fq.stopCh case is "picked" that we still return the error for the context

joestringer replied:
o_o;;;  OK yeah that would be more deterministic. I'll fix it up.

joestringer committed:
Add ResumableHash with sha1 implementation

This is mostly a convenience wrapper around the standard library which
will be used by the upcoming templating work.

Signed-off-by: Joe Stringer <joe@cilium.io>

borkmann reviewed (approved):
lgtm, nice work!

joestringer committed:
loader: Quieten symbol warning messages

Depending on whether ipv4/ipv6 are enabled/disabled, the symbol
substitutions in the ELF files may not be made for the corresponding
addresses. In these cases, rather than printing the following warnings
in the logs when skipping overwrite of the IPs, skip the warning:

    level=warning msg="Skipping symbol substitution" subsys=elf symbol=LXC_IP_1

Signed-off-by: Joe Stringer <joe@cilium.io>

tgraf commented on pkg/datapath/loader/cache.go:
```diff
@@ -72,6 +109,113 @@ func (o *objectCache) Update(nodeCfg *datapath.LocalNodeConfiguration) {
+	objectPath := filepath.Join(templatePath, endpointObj)
+
+	if err := os.MkdirAll(templatePath, defaults.StateDirRights); err != nil {
+		return err
```
Please embed the operation performed and value for templatePath into the error as debugging the error will be very difficult otherwise

tgraf commented on pkg/datapath/loader/cache.go:
```diff
@@ -72,6 +109,113 @@ func (o *objectCache) Update(nodeCfg *datapath.LocalNodeConfiguration) {
+	// Fetch the result of the compilation.
+	path, ok := o.lookup(hash)
+	if !ok {
+		return "", false, fmt.Errorf("peer compilation for this template failed")
```
I would not be sure what this error means if I saw it

joestringer replied:
Any thoughts on a better way to express that "Another goroutine tried to compile but failed with this error."? (In retrospect maybe these exact words would be more clear... ;) )

tgraf commented on pkg/datapath/loader/loader.go:
```diff
@@ -97,6 +98,40 @@ func CompileAndLoad(ctx context.Context, ep endpoint) error {
+	dstPath := path.Join(ep.StateDir(), endpointObj)
+	opts, strings := ELFSubstitutions(ep)
+	if err = template.Write(dstPath, opts, strings); err != nil {
+		return err
```
Does the error returned contain the `dstPath`?

joestringer replied:
Yes, the `pkg/elf` library puts the path into the error.

tgraf commented on pkg/serializer/func_queue.go:
```diff
@@ -81,6 +85,17 @@ func (fq *FunctionQueue) Stop() {
+	case <-fq.stopCh:
+	case <-ctx.Done():
+	}
+	return ctx.Err()
```
I'm assuming that this will return a generic `context deadline exceeded`? It would be helpful to return a more specific error so we know where it is coming from

joestringer replied:
I think this is up to the caller. If the context isn't exceeded, then this returns nil, which is the "Done" case.

joestringer replied:
Changed my mind, new patch on top.

tgraf commented on pkg/datapath/loader/cache.go:
```diff
@@ -65,6 +65,18 @@ func Init(dp datapath.Datapath, nodeCfg *datapath.LocalNodeConfiguration) {
+	if os.IsNotExist(err) {
+		return nil
+	}
+	return err
```
Return a more specific error message here

tgraf reviewed (changes requested):
This is amazing, only small nits around errors being returned

joestringer committed:
loader: Improve error messages

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
elf: Improve error messages

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
serializer: Add simple context to Wait() error msg

Signed-off-by: Joe Stringer <joe@cilium.io>
